src/sentry/workflow_engine/endpoints/validators/metric_alert_detector.py

```diff
@@ -38,3 +43,75 @@ def validate(self, attrs):
+
+    def update_data_conditions(self, instance, data_conditions):
+        """
+        Update the data condition if it already exists, create one if it does not
+        """
+        if instance.workflow_condition_group:
+            try:
+                data_condition_group = DataConditionGroup.objects.get(
+                    id=instance.workflow_condition_group.id
+                )
+            except DataConditionGroup.DoesNotExist:
+                raise serializers.ValidationError("DataConditionGroup not found, can't update")
+        # TODO make one if it doesn't exist and data is passed?
+
+        for data_condition in data_conditions:
+            try:
+                current_data_condition = DataCondition.objects.get(
+                    id=data_condition.get("id"), condition_group=data_condition_group
+                )
+            except DataCondition.DoesNotExist:
+                current_data_condition = None
+
+            if current_data_condition:
+                current_data_condition.update(**data_condition)
+            else:
+                DataCondition.objects.create(
+                    type=data_condition.get("type"),
+                    comparison=data_condition.get("comparison"),
+                    condition_result=data_condition.get("condition_result"),
+                    condition_group=data_condition_group,
+                )
+        return data_condition_group
+
+    def update_data_source(self, instance, data_source):
+        try:
+            source_instance = DataSource.objects.get(detector=instance)
+        except DataSource.DoesNotExist:
+            return
+        if source_instance:
+            try:
+                snuba_query = SnubaQuery.objects.get(id=source_instance.source_id)
+            except SnubaQuery.DoesNotExist:
+                raise serializers.ValidationError("SnubaQuery not found, can't update")
+
+        event_types = SnubaQueryEventType.objects.filter(snuba_query_id=snuba_query.id)
+        update_snuba_query(
+            snuba_query=snuba_query,
+            query_type=data_source.get("query_type", snuba_query.type),
+            dataset=data_source.get("dataset", snuba_query.dataset),
+            query=data_source.get("query", snuba_query.query),
+            aggregate=data_source.get("aggregate", snuba_query.aggregate),
+            time_window=timedelta(minutes=data_source.get("time_window", snuba_query.time_window)),
+            resolution=timedelta(seconds=data_source.get("resolution", snuba_query.resolution)),
+            environment=data_source.get("environment", snuba_query.environment),
+            event_types=data_source.get("event_types", [event_types]),
+        )
+
+    def update(self, instance, validated_data):
+        instance.name = validated_data.get("name", instance.name)
+        instance.type = validated_data.get("detector_type", instance.group_type).slug
+        condition_group = validated_data.pop("condition_group")
```

Comment:
nit: is there a reason this is `pop` instead of `get`?

Reply:
🤔 I wanna say I had a reason at one point but I don't think it applies anymore

Reply:
oh wait the reason is because when I call `instance.save()` the instance here is a `Detector` and if I have the dict of condition group data that's bad cause it's not data that gets saved to that model

Reply:
🤔 should this be the [`Detector.workflow_condition_group`](https://github.com/getsentry/sentry/blob/master/src/sentry/workflow_engine/models/detector.py#L49-L55)?

Reply:
What gets passed here are the data conditions, the condition group doesn't change